src/test/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitCheckerTest.java

```diff
@@ -0,0 +1,643 @@
+     *
+     * @author  Alex Taylor
+     */
+    @Test
+    public void ExpectedResetTimingNormalize() throws Exception {
+
+        // Use a longer notification interval to make the test produce stable output
+        ApiRateLimitChecker.setNotificationWaitMillis(2000);
+
+        // Set up scenarios
+        List<RateLimit> scenarios = new ArrayList<>();
+        int limit = 1000;
+        int buffer = 50;
+        scenarios.add(new RateLimit(limit, limit, soon));
+        //Giving a bit of time to make sure the setup happens on time
+        long start = System.currentTimeMillis() + 7000;
+
+        for (int i = 0; i <= 3; i++) {
+            scenarios.add(new RateLimit(limit, buffer - 5, new Date(start)));
+        }
+        // Refresh rate limit
+        scenarios.add(new RateLimit(limit, limit, soon));
+        setupStubs(scenarios);
+        // First server warm up
+        ApiRateLimitChecker.ThrottleForNormalize.checkApiRateLimit(listener, github);
+
+        while(System.currentTimeMillis() + 5500 < start)
+        {
+            Thread.sleep(25);
+        }
+
+        // Run check a few times to ensure we don't get stuck
+        ApiRateLimitChecker.ThrottleForNormalize.checkApiRateLimit(listener, github);
+
+        // Expect a triggered throttle for normalize
+        assertEquals(2, countOfOutputLinesContaining("Current quota"));
+        assertEquals(2, countOfOutputLinesContaining("Still sleeping"));
+    }
+
+    /**
+     * Verify the expected reset happens and notifications happen on time in the logs for OnOver
+     *
+     * @author  Alex Taylor
+     */
+    @Test
+    public void ExpectedResetTimingOnOver() throws Exception {
+
+        // Use a longer notification interval to make the test produce stable output
+        ApiRateLimitChecker.setNotificationWaitMillis(2000);
+
+        // Set up scenarios
+        List<RateLimit> scenarios = new ArrayList<>();
+        int limit = 1000;
+        int buffer = 50;
+        scenarios.add(new RateLimit(limit, limit, soon));
+        //Giving a bit of time to make sure the setup happens on time
+        long start = System.currentTimeMillis() + 7000;
+
+        for (int i = 0; i <= 3; i++) {
+            scenarios.add(new RateLimit(limit, buffer - 5, new Date(start)));
+        }
+        // Refresh rate limit
+        scenarios.add(new RateLimit(limit, limit, soon));
+        setupStubs(scenarios);
+        // First server warm up
+        ApiRateLimitChecker.ThrottleOnOver.checkApiRateLimit(listener, github);
+
+        while(System.currentTimeMillis() + 5500 < start)
+        {
+            Thread.sleep(25);
+        }
+
+        ApiRateLimitChecker.ThrottleOnOver.checkApiRateLimit(listener, github);
+
+        // We have 7 "notify" type messages and 2 "expired" type messages
+        assertEquals(2, countOfOutputLinesContaining("Current quota"));
+        assertEquals(2, countOfOutputLinesContaining("Still sleeping"));
+    }
+}
```

Comment:
nit: I'd add a newline here. The next person that changes the file will probably add one inadvertently due to editor settings anyway.
```suggestion
}

```